compiler/src/dmd/mtype.d

```diff
@@ -7343,3 +7192,157 @@ private extern(D) bool isCopyConstructorCallable (StructDeclaration argStruct,
     }
     return false;
 }
+
+/**
+ * Given a `Parameter` `p` and an argument `arg`, return the `MATCH` level.
+ *
+ * Params:
+ *   tf = The `TypeFunction` in which this `Parameter` lives, only used for
+ *        generating error messages with `pMessage`
+ *   p = The `Parameter` that is being matched
+ *   arg = The `Expression` given as an argument to `p`
+ *   wildmatch = The `inout` modifier
+ *   flag = Whether to check for the value implicit convertible-ness
+ *   sc = The `Scope` to use
+ *   pMessage = A pointer to store an error message in (can be `null`)
+ */
+private extern(D) MATCH matchParameter (TypeFunction tf, Parameter p, Expression arg,
+    ubyte wildmatch, bool flag, Scope* sc, const(char)** pMessage)
+{
+    assert(arg);
+    //printf("arg: %s, type: %s\n", arg.toChars(), arg.type.toChars());
+
+    MATCH m;
+    Type targ = arg.type;
+    Type tprm = wildmatch ? p.type.substWildTo(wildmatch) : p.type;
+
+    if (p.isLazy() && tprm.ty == Tvoid && targ.ty != Tvoid)
+        m = MATCH.convert;
+    else if (flag)
+    {
+        // for partial ordering, value is an irrelevant mockup, just look at the type
+        m = targ.implicitConvTo(tprm);
+    }
+    else
+    {
+        const isRef = p.isReference();
+        StructDeclaration argStruct, prmStruct;
+
+        // first look for a copy constructor
+        if (arg.isLvalue() && !isRef && targ.ty == Tstruct && tprm.ty == Tstruct)
+        {
+            // if the argument and the parameter are of the same unqualified struct type
+            argStruct = (cast(TypeStruct)targ).sym;
+            prmStruct = (cast(TypeStruct)tprm).sym;
+        }
+
+        // check if the copy constructor may be called to copy the argument
+        if (argStruct && argStruct == prmStruct && argStruct.hasCopyCtor)
+        {
+            if (!isCopyConstructorCallable(argStruct, arg, tprm, sc, pMessage))
+                return MATCH.nomatch;
+            m = MATCH.exact;
+        }
+
+        import dmd.dcast : cimplicitConvTo;
+        m = (sc && sc.flags & SCOPE.Cfile) ? arg.cimplicitConvTo(tprm) : arg.implicitConvTo(tprm);
+    }
+
+    // Non-lvalues do not match ref or out parameters
+    if (p.isReference())
```

Comment:
could use an early return here.

Reply:
I don't think an early return makes the code clearer here. If anything, the reference code could be separated in its own function, which IMO would be better. But keep in mind I'm doing those refactors to make it easier to implement named parameters.